cpp/src/rolling/detail/rolling_utils.cuh

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright (c) 2024-2025, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/detail/iterator.cuh>
+#include <cudf/types.hpp>
+
+#include <cuda/functional>
+
+namespace cudf {
+
+namespace detail::rolling {
+
+/**
+ * @brief A group descriptor for an ungrouped rolling window.
+ *
+ * @param num_rows The number of rows to be rolled over.
+ *
+ * @note This is used for uniformity of interface between grouped and ungrouped iterator
+ * construction.
+ */
+struct ungrouped {
+  cudf::size_type num_rows;
+
+  [[nodiscard]] __device__ constexpr cudf::size_type label(cudf::size_type) const noexcept
+  {
+    return 0;
+  }
+  [[nodiscard]] __device__ constexpr cudf::size_type start(cudf::size_type) const noexcept
+  {
+    return 0;
+  }
+  [[nodiscard]] __device__ constexpr cudf::size_type end(cudf::size_type) const noexcept
+  {
+    return num_rows;
+  }
+};
+
+/**
+ * @brief A group descriptor for a grouped rolling window.
+ *
+ * @param labels The group labels, mapping from input rows to group.
+ * @param offsets The group offsets providing the endpoints of each group.
+ *
+ * @note This is used for uniformity of interface between grouped and ungrouped iterator
+ * construction.
+ */
+struct grouped {
+  cudf::device_span<size_type const> const& labels;
+  cudf::device_span<size_type const> const& offsets;
```

Comment:
Spans don't really need to be taken in by reference

```suggestion
  cudf::device_span<size_type const> const labels;
  cudf::device_span<size_type const> const offsets;
```

Reply:
Thinking about efficiency/register pressure, each span costs 2 registers (pointer + size). So 2 spans cost 4 registers. We can do better by using just 3 registers instead: passing by 2 pointers + 1 size parameter.

Reply:
Right, but libcudf generally eschews raw ptrs.

Reply:
fwiw, I never need the size either because it's encoded in the for_each

Reply:
Then you can save 2 registers :smile:. Typically, this optimization is overkill. However, sometimes it works really well if the kernel is strained by a lot of computation with heavy register usage.

Reply:
I've gone back in the end to raw pointers for this reason, with a comment explaining why.